WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ShipmentProvidersViewController.swift

```diff
@@ -155,34 +160,13 @@ private extension ShipmentProvidersViewController {
     /// Registers for all of the related Notifications
     ///
     func startListeningToNotifications() {
-        let nc = NotificationCenter.default
-        nc.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
-        nc.addObserver(self, selector: #selector(keyboardWillHide), name: UIResponder.keyboardWillHideNotification, object: nil)
-    }
-
-    /// Executed whenever `UIResponder.keyboardWillShowNotification` note is posted
-    ///
-    @objc func keyboardWillShow(_ note: Notification) {
-        let bottomInset = keyboardHeight(from: note)
-
-        table.contentInset.bottom = bottomInset
-        table.scrollIndicatorInsets.bottom = bottomInset
+        keyboardFrameObserver.startObservingKeyboardFrame()
     }
 
-    /// Executed whenever `UIResponder.keyboardWillhideNotification` note is posted
-    ///
-    @objc func keyboardWillHide(_ note: Notification) {
-        table.contentInset.bottom = .zero
-        table.scrollIndicatorInsets.bottom = .zero
-    }
-
-    /// Returns the Keyboard Height from a (hopefully) Keyboard Notification.
-    ///
-    func keyboardHeight(from note: Notification) -> CGFloat {
-        let wrappedRect = note.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue
-        let keyboardRect = wrappedRect?.cgRectValue ?? .zero
-
-        return keyboardRect.height
+    func handleKeyboardFrameUpdate(keyboardFrame: CGRect) {
```

Comment:
This might be a bit overkill now, but given that all VCs that need to respond to keyboard are going to need to implement this exact code, and there might be more of those in the nearly future, have you considered factoring this out into an extension?

That would require the `table` property to be exposed, which in turn might exclude potential collection views, so I was thinking (and this just a mental exercise at this point, some architectural gymnastics if you will):

Let's assume we had a protocol that exposed a UIScrollView:

```
protocol KeyboardScrollable {
	var scrollable: UIScrollView { get }

	func handleKeyboardFrameUpdate(keyboardFrame: CGRect)
}

extension KeyboardScrollable {
	func handleKeyboardFrameUpdate(keyboardFrame: CGRect) {
		let keyboardHeight = keyboardFrame.height
		scrollable.contentInset.bottom = keyboardHeight
		scrollable.scrollIndicatorInsets.bottom = keyboardHeight
	}
}

extension ShipmentProvidersViewController: KeyboardScrollable {
	var scrollable: UIScrollView {
		return table
	}
}

extension WhatNotViewController: KeyboardScrollable {
	var scrollable: UIScrollView {
		return whatnot
	}	
}
```

it's not a huge win, but it removes code from view controllers, and still allows customisation if needed:

```
extension ViewControllerThatRequiresCustomisation: KeyboardScrollable {
	var scrollable: UIScrollView {
		return aCollectionView
	}

	func handleKeyboardFrameUpdate(keyboardFrame: CGRect) {
		// For some reason, we do not want to scroll here
		scrollable.contentInset.bottom = 0
	}

}
```

Reply:
I like this way of taking the scroll view inset adjustment to a protocol extension! Implemented in https://github.com/woocommerce/woocommerce-ios/pull/1430/commits/251328372bbbc72cfefdb299c251656c1f9dd285

it's not ideal for the scroll view to be exposed, I wish there's a way to privately implement a protocol with private `var` properties